### DISCUSSION TITLE
Flatten data when using batch data loaders

### DISCUSSION THREAD

**Discussion by lmeadors** [graphql-java/graphql-java#2324] [Q&A]
I'm building a graphql layer to aggregate 4 REST services:
- order
- product details
- availability
- pricing

I've started using `MappedBatchLoaderWithContext` classes to handle the service calls and that's working well - I get an incoming request for an order, fetch it with the order lines, then do batched requests to the 3 other rest services (product details, availability, and pricing services). 

All 3 of the other requests are batches and run in parallel - that's pretty amazing. :)

The response shape currently looks something like this:

```json
{
    "id": "123456",
    "price": 65.0,
    "qty": 3,
    "subtotal": 195.0,
    "product": {
        "productId": "234567",
        "name": "My Product",
        "price": {
            "price": 65.0,
            "retailPrice": 65.0,
            "savingsAmount": 0.0,
            "priceType": "RETAIL"
        },
        "detail": {
            "aka1": null,
            "aka2": null,
            "aka3": null
        },
        "availability": {
            "current": 123, 
            "blocks": 234
        }
    }
}
```

I would like to get it to look like this:

```json
    {
        "id": "123456",
        "price": 65.0,
        "qty": 3,
        "subtotal": 195.0,
        "product": {
            "productId": "234567",
            "name": "My Product",
            "price": 65.0,
            "retailPrice": 65.0,
            "savingsAmount": 0.0,
            "priceType": "RETAIL",
            "aka1": null,
            "aka2": null,
            "aka3": null,
            "current": 123, 
            "blocks": 234
        }
    }
```

Essentially what I'd like to do is take the child objects (`"price"`, `"detail"`, and `"availability"`) and merge their properties up to the `"product"` object, then discard them.

I read the data mapping doc on the site, and read "take these 3 sources of information and present them as one unified type", and thought "YES, that's exactly what I want!"

However, the two options presented there were both not really compatible with what I'm doing - I don't have (or want) a fetcher that gets all of the data and pulls it together into a single uber bean - I'd like to instead do it at runtime, essentially making a union of the 4 objects (`"product"`, `"price"`, `"detail"`, and `"availability"`) and calling the result of that my `"product"`.

I'm sure there is a way to do this, but I can't seem to see it - any advice?

**Comment by aleksandarsusnjar** [comment]
I didn't read the documentation you are mentioning but I am refactoring my data in related manner. I don't construct "uber beans" fully... but I do construct a proxy that I can return. I have a feeling that this proxy can even be a singleton because you could play with the "local context" and get stuff from there anyway... but I also feel that'd be wrong. I do hate pointless data copying so I would not stuff the "uber bean" with any data... It would just be a "flattening proxy".
